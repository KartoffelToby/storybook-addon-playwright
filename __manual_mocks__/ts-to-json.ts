export const createProgramMock = jest.fn();
export const createParserMock = jest.fn();
export const createFormatterMock = jest.fn();
export const SchemaGeneratorMock = jest.fn().mockImplementation(() => {
  return {
    createSchema: () => ({
      definitions: { ['MyType']: { properties: { props: true } } },
    }),
  };
});
jest.mock('ts-to-json', () => ({
  SchemaGenerator: SchemaGeneratorMock,
  createFormatter: createFormatterMock,
  createParser: createParserMock,
  createProgram: createProgramMock,
}));
