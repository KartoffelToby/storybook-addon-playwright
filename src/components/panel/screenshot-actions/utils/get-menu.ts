import { Definition } from 'ts-to-json';
import { ActionMenuItemBase } from '../ActionMenuItem';

export const getMenu = (
  def: Definition,
  parents: string[] = [],
): ActionMenuItemBase[] => {
  const actionArr = Object.keys(def);
  let menuItems: ActionMenuItemBase[] = [];
  actionArr.forEach((actionKey) => {
    const prop = def[actionKey];

    if (prop.type === 'object') {
      const childProps = getMenu(prop.properties, [...parents, actionKey]);
      menuItems = [...menuItems, ...childProps];
    } else {
      const label = [...parents, actionKey];
      menuItems.push({
        label: label.join('.'),
        name: label.join('.'),
      });
    }
  });

  return menuItems;
};
