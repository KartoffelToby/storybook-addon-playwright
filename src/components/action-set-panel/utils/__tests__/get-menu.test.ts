import { getMenu } from '../get-menu';
import { Definition } from 'ts-to-json';
const definition = {
  click: {
    kind: 'function',
    type: 'Promise',
  },
  mouse: {
    additionalProperties: false,
    properties: {
      click: {
        description: 'Shortcut for `mouse.move`, `mouse.down` and `mouse.up`.',
        kind: 'function',

        required: ['x', 'y'],
        type: 'Promise',
      },
      dblclick: {
        kind: 'function',
        required: ['x', 'y'],
        type: 'Promise',
      },
    },
    required: ['click', 'dblclick'],
    type: 'object',
  },
} as Definition;

describe('getMenu', () => {
  it('should return valid', () => {
    expect(getMenu(definition)).toMatchSnapshot();
  });
});
