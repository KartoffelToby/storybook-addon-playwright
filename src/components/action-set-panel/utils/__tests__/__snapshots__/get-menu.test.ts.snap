// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getMenu should return valid 1`] = `
Array [
  Object {
    "label": "click",
    "name": "click",
  },
  Object {
    "label": "mouse.click",
    "name": "mouse.click",
  },
  Object {
    "label": "mouse.dblclick",
    "name": "mouse.dblclick",
  },
]
`;
